pkg/agent/client/experimental.go

```diff
@@ -0,0 +1,357 @@
+	if err != nil {
+		return nil, err
+	}
+	defer connection.Release()
+
+	stream, err := agentClient.RenewAgent(ctx)
+	if err != nil {
+		c.release(connection)
+		c.c.Log.WithError(err).Error("failed to renew connection")
+		return nil, err
+	}
+
+	req := &agentpb.RenewAgentRequest{
+		Step: &agentpb.RenewAgentRequest_Params{
+			Params: &agentpb.AgentX509SVIDParams{
+				Csr: csr,
+			},
+		},
+	}
+	if err := stream.Send(req); err != nil {
+		return nil, err
+	}
+
+	resp, err := stream.Recv()
+	if err != nil {
+		return nil, err
+	}
+
+	req = &agentpb.RenewAgentRequest{
+		Step: &agentpb.RenewAgentRequest_Ack_{Ack: &agentpb.RenewAgentRequest_Ack{}},
+	}
+	if err := stream.Send(req); err != nil {
+		return nil, err
+	}
+	if _, err := stream.Recv(); err != io.EOF {
+		return nil, err
+	}
+
+	return resp.Svid, nil
+}
+
+func (c *client) fetchSVIDs(ctx context.Context, params []*svidpb.NewX509SVIDParams) ([]*types.X509SVID, error) {
+	svidClient, connection, err := c.newSVIDClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer connection.Release()
+
+	resp, err := svidClient.BatchNewX509SVID(ctx, &svidpb.BatchNewX509SVIDRequest{
+		Params: params,
+	})
+	if err != nil {
+		c.release(connection)
+		c.c.Log.WithError(err).Error("failed to batch svid")
+		return nil, errors.New("failed to batch svid")
+	}
+
+	okStatus := int32(codes.OK)
+	var svids []*types.X509SVID
+	for i, r := range resp.Results {
+		if r.Status.Code != okStatus {
+			c.c.Log.WithFields(logrus.Fields{
+				telemetry.RegistrationID: params[i].EntryId,
+				telemetry.Status:         r.Status.Code,
+				telemetry.Error:          r.Status.Message,
+			}).Warn("Fails to mint X509 SVID")
+		}
+
+		svids = append(svids, r.Bundle)
+	}
+
+	return svids, nil
+}
+
+func (c *client) fetchJWTSVID(ctx context.Context, jsr *node.JSR, entryID string) (*JWTSVID, error) {
+	svidClient, connection, err := c.newSVIDClient(ctx)
+	if err != nil {
+		return nil, err
+	}
+	defer connection.Release()
+
+	resp, err := svidClient.NewJWTSVID(ctx, &svidpb.NewJWTSVIDRequest{
+		Audience: jsr.Audience,
+		EntryId:  entryID,
+	})
+	if err != nil {
+		c.release(connection)
+		c.c.Log.WithError(err).Errorf("Failure fetching JWT SVID")
+		return nil, fmt.Errorf("failure fetching JWT SVID: %v", err)
+	}
+
+	svid := resp.Svid
+	switch {
+	case svid == nil:
+		return nil, errors.New("JWTSVID response missing SVID")
+	case svid.IssuedAt == 0:
+		return nil, errors.New("JWTSVID missing issued at")
+	case svid.ExpiresAt == 0:
+		return nil, errors.New("JWTSVID missing expires at")
+	case svid.IssuedAt > svid.ExpiresAt:
+		return nil, errors.New("JWTSVID issued after it has expired")
+	}
+
+	return &JWTSVID{
+		Token:     svid.Token,
+		IssuedAt:  time.Unix(svid.IssuedAt, 0).UTC(),
+		ExpiresAt: time.Unix(svid.ExpiresAt, 0).UTC(),
+	}, nil
+}
+
+func (c *client) newEntryClient(ctx context.Context) (entrypb.EntryClient, *nodeConn, error) {
+	c.m.Lock()
+	defer c.m.Unlock()
+
+	// open a new connection
```

Comment:
nit: not sure if we really need to have this comment. In case that we have it, maybe "check if we need to open a new connection" is more appropriate. Or move this inside.
Same with other similar comments.